azure/services/natgateways/natgateways.go

```diff
@@ -18,156 +18,112 @@ package natgateways
+	var resultingErr error
 	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
-		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
-
-		switch {
-		case err != nil && !azure.ResourceNotFound(err):
-			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
-		case err == nil:
-			// nat gateway already exists
-			log.V(4).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
-			natGatewaySpec.Subnet.NatGateway.ID = existingNatGateway.ID
-
-			if existingNatGateway.NatGatewayIP.Name == natGatewaySpec.NatGatewayIP.Name {
-				// Skip update for Nat Gateway as it exists with expected values
-				log.V(4).Info("Nat Gateway exists with expected values, skipping update", "nat gateway", natGatewaySpec.Name)
-				natGatewaySpec.Subnet.NatGateway = *existingNatGateway
-				s.Scope.SetSubnet(natGatewaySpec.Subnet)
-				continue
-			} else {
-				log.V(2).Info("updating NAT gateway IP name to match the spec", "old name", existingNatGateway.NatGatewayIP.Name, "desired name", natGatewaySpec.NatGatewayIP.Name)
+		result, err := s.CreateResource(ctx, natGatewaySpec, serviceName)
+		if err != nil {
+			if !azure.IsOperationNotDoneError(err) || resultingErr == nil {
+				resultingErr = err
 			}
-		default:
-			// nat gateway doesn't exist but its name was specified in the subnet, let's create it
-			log.V(2).Info("nat gateway doesn't exist yet, creating it", "nat gateway", natGatewaySpec.Name)
 		}
+		if err == nil {
+			natGateway, ok := result.(network.NatGateway)
+			if !ok {
+				// Return out of loop since this would be an unexpcted fatal error
+				resultingErr = errors.Errorf("created resource %T is not a network.NatGateway", result)
+				break
+			}
 
-		natGatewayToCreate := network.NatGateway{
-			Location: to.StringPtr(s.Scope.Location()),
-			Sku:      &network.NatGatewaySku{Name: network.Standard},
-			NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
-				PublicIPAddresses: &[]network.SubResource{
-					{
-						ID: to.StringPtr(azure.PublicIPID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.NatGatewayIP.Name)),
-					},
-				},
-			},
-		}
-		err = s.client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), natGatewaySpec.Name, natGatewayToCreate)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create nat gateway %s in resource group %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
-		}
-		log.V(2).Info("successfully created nat gateway", "nat gateway", natGatewaySpec.Name)
-		natGateway := infrav1.NatGateway{
-			ID:   azure.NatGatewayID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.Name),
-			Name: natGatewaySpec.Name,
-			NatGatewayIP: infrav1.PublicIPSpec{
-				Name: natGatewaySpec.NatGatewayIP.Name,
-			},
+			// TODO: ideally we wouldn't need to set the subnet spec based on the result of the create operation
+			s.Scope.SetNatGatewayIDInSubnets(natGatewaySpec.ResourceName(), *natGateway.ID)
```

Comment:
can you please add a comment about this one in #1685 so we remember it when we fix the issue